tests/framework/microvm.py

```diff
@@ -926,6 +926,14 @@ def ssh(self):
         """Return a cached SSH connection on the 1st interface"""
         return self.ssh_iface(0)
 
+    def wait_for_up(self, timeout=10):
+        """Wait for guest running inside the microcode to come up and respond.
+
+        :param timeout: seconds to wait.
+        """
+        rc, _, stderr = self.ssh.run("true", timeout)
+        assert rc == 0, stderr
```

Comment:
Proposal: now that you have it centralized, we could expand that message:

```
Command did not respond within {timeout}s

STDOUT={stdout}

STDERR={stderr}

# and maybe?:
Firecracker logs: {self.log_data}

Thread backtrace:
for threads in self.thread:
    cat /proc/thread_pid/stack
```

But I understand if you want this as a follow-up

Reply:
I'll do that! thanks for your great suggestion!

Reply:
I think this would be nice to have inside `self.ssh.run()` method

Reply:
when timeout happens, `subprocess.TimeoutExpired` is raised in `utils.run_cmd_sync()`. so let's skip `Command did not respond within {timeout}s` here.

Reply:
Does this look good to you?
```
Remote command did not respond within 0.1s

Firecracker logs:
2024-04-19T11:00:44.089244919 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] Running Firecracker v1.8.0-dev
2024-04-19T11:00:44.290029494 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The API server received a Put request on "/machine-config" with body "{\"vcpu_count\": 2, \"mem_size_mib\": 512, \"track_dirty_pages\": false}".
2024-04-19T11:00:44.290259648 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The request was executed successfully. Status code: 204 No Content.
2024-04-19T11:00:44.290265998 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] Total previous API call duration: 265 us.
2024-04-19T11:00:44.291760062 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The API server received a Put request on "/boot-source" with body "{\"kernel_image_path\": \"/vmlinux-4.14.336\"}".
2024-04-19T11:00:44.291783118 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The request was executed successfully. Status code: 204 No Content.
2024-04-19T11:00:44.291786455 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] Total previous API call duration: 29 us.
2024-04-19T11:00:44.292843385 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The API server received a Put request on "/drives/rootfs" with body "{\"drive_id\": \"rootfs\", \"path_on_host\": \"/ubuntu-22.04.squashfs\", \"is_root_device\": true, \"is_read_only\": true}".
2024-04-19T11:00:44.292900872 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The request was executed successfully. Status code: 204 No Content.
2024-04-19T11:00:44.292904918 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] Total previous API call duration: 65 us.
2024-04-19T11:00:44.578897387 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The API server received a Put request on "/network-interfaces/eth0" with body "{\"iface_id\": \"eth0\", \"host_dev_name\": \"tap0\", \"guest_mac\": \"06:00:c0:a8:00:02\"}".
2024-04-19T11:00:44.579656955 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The request was executed successfully. Status code: 204 No Content.
2024-04-19T11:00:44.579670421 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] Total previous API call duration: 778 us.
2024-04-19T11:00:44.580240045 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The API server received a Put request on "/vsock" with body "{\"vsock_id\": \"vsock0\", \"guest_cid\": 3, \"uds_path\": \"v.sock\"}".
2024-04-19T11:00:44.580303140 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The request was executed successfully. Status code: 204 No Content.
2024-04-19T11:00:44.580306720 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] PUT /vsock: vsock_id field is deprecated.
2024-04-19T11:00:44.580309499 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] Total previous API call duration: 73 us.
2024-04-19T11:00:44.580841372 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The API server received a Get request on "/".
2024-04-19T11:00:44.580854446 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The request was executed successfully. Status code: 200 OK.
2024-04-19T11:00:44.580860383 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] Total previous API call duration: 20 us.
2024-04-19T11:00:44.581588739 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The API server received a Put request on "/actions" with body "{\"action_type\": \"InstanceStart\"}".
2024-04-19T11:00:44.581601638 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] event_start: build microvm for boot
2024-04-19T11:00:44.587082940 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] acpi: Building AML for VirtIO device _SB_.V000. memory range: 0xd0000000:4096 irq: 5
2024-04-19T11:00:44.587122323 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] acpi: Building AML for VirtIO device _SB_.V001. memory range: 0xd0001000:4096 irq: 6
2024-04-19T11:00:44.587150759 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] acpi: Building AML for VirtIO device _SB_.V002. memory range: 0xd0002000:4096 irq: 7
2024-04-19T11:00:44.587156397 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] vmgenid: building VMGenID device. Address: 0x000df000. IRQ: 8
2024-04-19T11:00:44.587171998 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] vmgenid: writing new generation ID to guest: 0xee86da13c0ac21a08b47b16357623044
2024-04-19T11:00:44.587314181 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] mptable: Allocated 324 bytes for MPTable 2 vCPUs at address 0x0009fc00
2024-04-19T11:00:44.587357288 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] acpi: Wrote table (755 bytes) at address: 0x0009fd44
2024-04-19T11:00:44.587361203 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] acpi: Wrote table (276 bytes) at address: 0x000a0037
2024-04-19T11:00:44.587364551 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] acpi: Wrote table (72 bytes) at address: 0x000a014b
2024-04-19T11:00:44.587367550 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] acpi: Wrote table (52 bytes) at address: 0x000a0193
2024-04-19T11:00:44.587370440 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] acpi: Wrote RSDP (36 bytes) at address: 0x000e0000
2024-04-19T11:00:44.587794537 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] event_end: build microvm for boot
2024-04-19T11:00:44.587797988 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] event_start: boot microvm
2024-04-19T11:00:44.587800087 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] Artificially kick devices.
2024-04-19T11:00:44.587816902 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_vcpu 0] Received a VcpuEvent::Resume message with immediate_exit enabled. immediate_exit was disabled before proceeding
2024-04-19T11:00:44.587823252 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_vcpu 1] Received a VcpuEvent::Resume message with immediate_exit enabled. immediate_exit was disabled before proceeding
2024-04-19T11:00:44.587837035 [68b73387-d84c-4710-8122-b4ed8bf3ccff:main] event_end: boot microvm
2024-04-19T11:00:44.587862303 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The request was executed successfully. Status code: 204 No Content.
2024-04-19T11:00:44.587870818 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] Total previous API call duration: 6284 us.
2024-04-19T11:00:44.588379651 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The API server received a Get request on "/".
2024-04-19T11:00:44.588396542 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] The request was executed successfully. Status code: 200 OK.
2024-04-19T11:00:44.588401523 [68b73387-d84c-4710-8122-b4ed8bf3ccff:fc_api] Total previous API call duration: 23 us.

Thread backtraces:
firecracker (pid=136):
[<0>] ep_poll+0x34d/0x3b0
[<0>] do_epoll_wait+0xc7/0xe0
[<0>] do_compat_epoll_pwait.part.0+0xc/0x70
[<0>] __x64_sys_epoll_pwait+0x7f/0x130
[<0>] do_syscall_64+0x38/0x90
[<0>] entry_SYSCALL_64_after_hwframe+0x64/0xce

fc_api (pid=137):
[<0>] ep_poll+0x34d/0x3b0
[<0>] do_epoll_wait+0xc7/0xe0
[<0>] do_compat_epoll_pwait.part.0+0xc/0x70
[<0>] __x64_sys_epoll_pwait+0x7f/0x130
[<0>] do_syscall_64+0x38/0x90
[<0>] entry_SYSCALL_64_after_hwframe+0x64/0xce

fc_vcpu 0 (pid=145):
[<0>] kvm_vcpu_block+0x52/0xc0 [kvm]
[<0>] kvm_vcpu_halt+0x67/0x3e0 [kvm]
[<0>] vcpu_run+0x1df/0x250 [kvm]
[<0>] kvm_arch_vcpu_ioctl_run+0x182/0x4e0 [kvm]
[<0>] kvm_vcpu_ioctl+0x271/0x6b0 [kvm]
[<0>] __x64_sys_ioctl+0x87/0xc0
[<0>] do_syscall_64+0x38/0x90
[<0>] entry_SYSCALL_64_after_hwframe+0x64/0xce

fc_vcpu 1 (pid=146):
[<0>] kvm_vcpu_block+0x52/0xc0 [kvm]
[<0>] kvm_vcpu_halt+0x67/0x3e0 [kvm]
[<0>] vcpu_run+0x1df/0x250 [kvm]
[<0>] kvm_arch_vcpu_ioctl_run+0x182/0x4e0 [kvm]
[<0>] kvm_vcpu_ioctl+0x271/0x6b0 [kvm]
[<0>] __x64_sys_ioctl+0x87/0xc0
[<0>] do_syscall_64+0x38/0x90
[<0>] entry_SYSCALL_64_after_hwframe+0x64/0xce

---------------------------------------------------------------
```

Reply:
Looks great!